pkg/service/controller/ingressip/controller.go

```diff
@@ -0,0 +1,606 @@
+	if !ipExists {
+		service.Spec.ExternalIPs = append(service.Spec.ExternalIPs, ingressIP)
+		glog.V(5).Infof("Adding ip %v to service %v as an external ip", ingressIP, key)
+		return persistServiceSpec(ic.client, service)
+	}
+	return nil
+}
+
+// allocateIP attempts to allocate the requested ip, and if that is
+// not possible, allocates the next available address.
+func (ic *IngressIPController) allocateIP(requestedIP string) (net.IP, error) {
+	if len(requestedIP) == 0 {
+		// Specific ip not requested
+		return ic.ipAllocator.AllocateNext()
+	}
+	var ip net.IP
+	if ip = net.ParseIP(requestedIP); ip == nil {
+		// Invalid ip
+		return ic.ipAllocator.AllocateNext()
+	}
+	if err := ic.ipAllocator.Allocate(ip); err != nil {
+		// Unable to allocate requested ip
+		return ic.ipAllocator.AllocateNext()
+	}
+	// Allocated requested ip
+	return ip, nil
+}
+
+// release wraps the allocator's Release method to log any error it returns.
+func (ic *IngressIPController) release(ip *net.IP, key string) {
+	if err := ic.ipAllocator.Release(*ip); err != nil {
+		// Release from contiguous allocator should never return an error, but just in case...
+		utilruntime.HandleError(fmt.Errorf("Error releasing ip %v for service %v: %v", ip.String(), key, err))
+	}
+}
+
+func persistServiceSpec(client kclient.ServicesNamespacer, service *kapi.Service) error {
+	return persistService(client, service, false)
+}
+
+func persistServiceStatus(client kclient.ServicesNamespacer, service *kapi.Service) error {
+	return persistService(client, service, true)
+}
+
+func persistService(client kclient.ServicesNamespacer, service *kapi.Service, targetStatus bool) error {
+	backoff := wait.Backoff{
+		Steps:    clientRetryCount,
+		Duration: clientRetryInterval,
+	}
+	return wait.ExponentialBackoff(backoff, func() (bool, error) {
+		var err error
+		if targetStatus {
+			_, err = client.Services(service.Namespace).UpdateStatus(service)
+		} else {
+			_, err = client.Services(service.Namespace).Update(service)
+		}
+		switch {
+		case err == nil:
+			return true, nil
+		case kerrors.IsNotFound(err):
+			// If the service no longer exists, we don't want to recreate
+			// it. Just bail out so that we can process the delete, which
+			// we should soon be receiving if we haven't already.
+			glog.V(5).Infof("Not persisting update to service '%s/%s' that no longer exists: %v",
+				service.Namespace, service.Name, err)
+			return true, nil
+		case kerrors.IsConflict(err):
+			// TODO: Try to resolve the conflict if the change was
+			// unrelated to load balancer status. For now, just rely on
+			// the fact that we'll also process the update that caused the
+			// resource version to change.
+			glog.V(5).Infof("Not persisting update to service '%s/%s' that has been changed since we received it: %v",
+				service.Namespace, service.Name, err)
+			return true, nil
+		default:
+			utilruntime.HandleError(fmt.Errorf("Failed to persist updated LoadBalancerStatus to service '%s/%s': %v",
```

Comment:
Since you return the error you don't need to utilruntime.HandleError here.  Basically reserve HandleError for "i am eating this error".  The code that calls this is responsible for logging it.